sdks/java/extensions/sql/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -1,106 +1,154 @@
-    ColumnDef(list)
-    ( <COMMA> ColumnDef(list) )*
+    id = SimpleIdentifier()
+    value = Literal() {
+        list.add(id);
+        list.add(value);
+    }
+}
+
+SqlNodeList TableElementList() :
+{
+    final Span s;
+    final List<SqlNode> list = Lists.newArrayList();
+}
+{
+    <LPAREN> { s = span(); }
+    TableElement(list)
+    (
+        <COMMA> TableElement(list)
+    )*
     <RPAREN> {
-        return new SqlNodeList(list, pos.plus(getPos()));
+        return new SqlNodeList(list, s.end(this));
+    }
+}
+
+void TableElement(List<SqlNode> list) :
+{
+    final SqlIdentifier id;
+    final SqlDataTypeSpec type;
+    final boolean nullable;
+    SqlNode comment = null;
+    final Span s = Span.of();
+}
+{
+    id = SimpleIdentifier()
+    (
+        type = DataType()
+        (
+            <NULL> { nullable = true; }
+        |
+            <NOT> <NULL> { nullable = false; }
+        |
+            { nullable = true; }
+        )
+        [ <COMMENT> comment = StringLiteral() ]
+        {
+            list.add(
+                SqlDdlNodes.column(s.add(id).end(this), id,
+                    type.withNullable(nullable), comment));
+        }
+    |
+        { list.add(id); }
+    )
+|
+    id = SimpleIdentifier() {
+        list.add(id);
     }
 }
 
-/**
- * CREATE TABLE ( IF NOT EXISTS )?
- *   ( database_name '.' )? table_name ( '(' column_def ( ',' column_def )* ')'
- *   ( STORED AS INPUTFORMAT input_format_classname OUTPUTFORMAT output_format_classname )?
- *   LOCATION location_uri
- *   ( TBLPROPERTIES tbl_properties )?
- *   ( AS select_stmt )
- */
-SqlNode SqlCreateTable() :
-{
-    SqlParserPos pos;
-    SqlIdentifier tblName;
-    SqlNodeList fieldList;
+SqlCreate SqlCreateTable(Span s, boolean replace) :
+{
+    final boolean ifNotExists;
+    final SqlIdentifier id;
+    SqlNodeList tableElementList = null;
     SqlNode type = null;
     SqlNode comment = null;
     SqlNode location = null;
-    SqlNode tbl_properties = null;
-    SqlNode select = null;
+    SqlNode tblProperties = null;
+    SqlNode query = null;
 }
 {
-    <CREATE> { pos = getPos(); }
-    <TABLE>
-    tblName = CompoundIdentifier()
-    fieldList = ColumnDefinitionList()
-    <TYPE>
-    type = StringLiteral()
-    [
-    <COMMENT>
-    comment = StringLiteral()
-    ]
-    [
-    <LOCATION>
-    location = StringLiteral()
-    ]
-    [ <TBLPROPERTIES> tbl_properties = StringLiteral() ]
-    [ <AS> select = OrderedQueryOrExpr(ExprContext.ACCEPT_QUERY) ] {
-        return new SqlCreateTable(pos, tblName, fieldList, type, comment,
-        location, tbl_properties, select);
+    <TABLE> ifNotExists = IfNotExistsOpt() id = CompoundIdentifier()
+    [ tableElementList = TableElementList() ]
+    <TYPE> type = StringLiteral()
+    [ <COMMENT> comment = StringLiteral() ]
+    [ <LOCATION> location = StringLiteral() ]
+    [ <TBLPROPERTIES> tblProperties = StringLiteral() ]
+    [ <AS> query = OrderedQueryOrExpr(ExprContext.ACCEPT_QUERY) ]
```

Comment:
How about remove the `query` statement here as it's not implemented actually? IMO we should separate `CREATE TABLE XX AS SELECT ...` from standard CREATE DDL.

Reply:
Removing `AS` until it is implemented is a great idea! Thanks for the suggestion, I've done that.